CrossEntropyLoss triggers floating point exception

mikeb127 opened the pull request:
Adds cast to float to fix floating point exception

Fixes #73165

mikeb127 committed:
Draft checks added to MultiLabelMarginCriterion.cu

mikeb127 committed:
Remove tab and replace with space

mikeb127 committed:
Update MultiLabelMarginCriterion.cu

mikeb127 committed:
Fix CrossEntropyLoss triggers floating point exception

mikeb127 committed:
Update LossNLL.cpp

mikeb127 committed:
CrossEntropyLoss triggers floating point exception

Cast to float to avoid exception

mikeb127 committed:
Update MultiLabelMarginCriterion.cu

mikeb127 committed:
Update MultiLabelMarginCriterion.cu

mikeb127 commented:
Reviewers - Apologies for what may be a stupid question, but I'm new to contributing to this repo....

The tests above that are failing seem to have nothing to do with what I changed at all. Do I need to address these before you guys will review/accept?

jbschlosser commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -491,7 +491,7 @@ Tensor cross_entropy_loss_prob_target(
     switch (reduction) {
       case Reduction::Mean:
-        return -(input * target * weight_).sum() / (input.numel() / input.size(1));
+        return -(input * target * weight_).sum() / ((float) input.numel() / (float) input.size(1));
```
Please use C++-style casts (i.e. `static_cast<float>(...)`) over C-style.

mikeb127 replied:
Changes made and tests added. Apologies for the ridiculous number of commits, I have manage to absolutely butcher a simple change.

jbschlosser replied:
No worries at all! No change is simple in this codebase tbh :)

mikeb127 replied:
Thanks :). Anything else you need from me for this one? Do you want me to try and rebase the PR to get rid of these unrelated errors? Or are you happy to review as is?

mikeb127 replied:
Converted back to draft to fix failing tests

mikeb127 replied:
All tests now passing. Let me know what you think

jbschlosser reviewed (changes requested):
Thanks for the fix! Sorry for the review delay - I was out last week.

I'm not convinced this solves the problem of 0 in the denominator. Please add a test in `test/test_nn.py` ensuring that the case referenced in the issue causes no issue after the fix:
```python
input_1 = torch.rand([5, 0], dtype=torch.float32)
input_2 = torch.rand([5, 0], dtype=torch.float32)
torch.nn.CrossEntropyLoss()(input_1, input_2)
```

mikeb127 committed:
Update LossNLL.cpp

mikeb127 committed:
Update test_nn.py

mikeb127 committed:
Update test_nn.py

mikeb127 committed:
Explicitly force NaN

Original solution is architecture dependent, tests fail it as undefined behavior. This commit changes the original solution to force the NaN value

mikeb127 committed:
Remove stray tab

mikeb127 committed:
Remove trailing space

mikeb127 committed:
Check test error

mikeb127 committed:
Update test_nn.py

mikeb127 committed:
Merge branch 'pytorch:master' into cross-entropy-error

mikeb127 committed:
Remove unreachable code

jbschlosser reviewed (approved):
LGTM thanks!
